## Duplicated features

Often datasets contain one or more features that show the same values across all the observations. This means that both features are in essence identical. In addition, it is not unusual to introduce duplicated features after performing **one hot encoding** of categorical variables, particularly when using several highly cardinal variables.

Identifying and removing duplicated, and therefore redundant features, is an easy first step towards feature selection and more easily interpretable machine learning models.

Here I will demonstrate how to identify duplicated features using the Santander Customer Satisfaction dataset from Kaggle. 

There is no function in python and pandas to find duplicated columns. I will show 2 snippets of code, one that you can apply to small datasets, and a second snippet that you can use on larger datasets. The first piece of code, is computationally costly, so your computer might run out of memory.

**Note**
Finding duplicated features is a computationally costly operation in Python, therefore depending on the size of your dataset, you might not always be able to perform it.

In [ ]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split

## Removing duplicate features

In [ ]:
# load the Santander customer satisfaction dataset from Kaggle
# I load just a few rows for the demonstration
data = pd.read_csv('santander.csv', nrows=15000)
data.shape

(15000, 371)

In [ ]:
# check the presence of null data.
# The snippets below will be able to compare nan values between 2 columns,
# so in principle missing data are not a problem.
# in any case, we see that there are no missing data in this dataset

[col for col in data.columns if data[col].isnull().sum() > 0]

[]

### Important

In all feature selection procedures, it is good practice to select the features by examining only the training set. And this is to avoid overfit.

In [ ]:
# separate train and test sets
X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['TARGET'], axis=1),
    data['TARGET'],
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((10500, 370), (4500, 370))

Pandas has the function 'duplicated' that evaluates if the dataframe contains duplicated rows. We can use this function to check for duplicated columns if we transpose the dataframe first. By transposing the dataframe, we obtain a new dataframe where the columns are now rows, and with the 'duplicated' method we can go ahead an identify those that are duplicated. 

Once we identify them, we can remove the duplicated rows. See below.

### Code Snippet for small datasets

Using pandas transpose is computationally expensive, so the computer may run out of memory. That is why we can only use this code block on small datasets. How small will depend of your computer specifications.

In [ ]:
# transpose the dataframe, so that the columns are the rows of the new dataframe
data_t = X_train.T
data_t.head()

10439    9236    818    ...    9845     10799   2732 
ID                       20941.0  18583.0  1623.0  ...  19796.0  21653.0  5441.0
var3                         2.0      2.0     2.0  ...      2.0      2.0     2.0
var15                       23.0     39.0    22.0  ...     38.0     28.0    23.0
imp_ent_var16_ult1           0.0      0.0   150.0  ...      0.0      0.0     0.0
imp_op_var39_comer_ult1      0.0      0.0     0.0  ...      0.0      0.0     0.0

[5 rows x 10500 columns]

In [ ]:
# check if there are duplicated rows (the columns of the original dataframe)
# this is a computionally expensive operation, so it might take a while
# sum indicates how many rows are duplicated

data_t.duplicated().sum()

105

We can see that 105 columns / variables are duplicated. This means that 105 variables are identical to at least another variable within a dataset.

In [ ]:
# visualise the duplicated rows (the columns of the original dataframe)
data_t[data_t.duplicated()]

10439  9236   818    ...  9845   10799  2732 
ind_var2                         0.0    0.0    0.0  ...    0.0    0.0    0.0
ind_var13_medio_0                0.0    0.0    0.0  ...    0.0    0.0    0.0
ind_var13_medio                  0.0    0.0    0.0  ...    0.0    0.0    0.0
ind_var18_0                      0.0    0.0    0.0  ...    0.0    0.0    0.0
ind_var18                        0.0    0.0    0.0  ...    0.0    0.0    0.0
...                              ...    ...    ...  ...    ...    ...    ...
saldo_medio_var13_medio_hace2    0.0    0.0    0.0  ...    0.0    0.0    0.0
saldo_medio_var13_medio_hace3    0.0    0.0    0.0  ...    0.0    0.0    0.0
saldo_medio_var13_medio_ult1     0.0    0.0    0.0  ...    0.0    0.0    0.0
saldo_medio_var13_medio_ult3     0.0    0.0    0.0  ...    0.0    0.0    0.0
saldo_medio_var29_hace3          0.0    0.0    0.0  ...    0.0    0.0    0.0

[105 rows x 10500 columns]

In [ ]:
# we can capture the duplicated features, by capturing the
# index values of the transposed dataframe like this:
duplicated_features = data_t[data_t.duplicated()].index.values
duplicated_features

array(['ind_var2', 'ind_var13_medio_0', 'ind_var13_medio', 'ind_var18_0',
       'ind_var18', 'ind_var26', 'ind_var25', 'ind_var27_0',
       'ind_var28_0', 'ind_var28', 'ind_var27', 'ind_var29_0',
       'ind_var29', 'ind_var32', 'ind_var34_0', 'ind_var34', 'ind_var37',
       'ind_var40_0', 'ind_var40', 'ind_var41', 'ind_var39', 'ind_var44',
       'ind_var46_0', 'ind_var46', 'num_var13_medio_0', 'num_var13_medio',
       'num_var18_0', 'num_var18', 'num_var26', 'num_var25',
       'num_var27_0', 'num_var28_0', 'num_var28', 'num_var27',
       'num_var29_0', 'num_var29', 'num_var32', 'num_var34_0',
       'num_var34', 'num_var37', 'num_var40_0', 'num_var40', 'num_var41',
       'num_var39', 'num_var46_0', 'num_var46', 'saldo_var13_medio',
       'saldo_var18', 'saldo_var28', 'saldo_var27', 'saldo_var29',
       'saldo_var34', 'saldo_var40', 'saldo_var41', 'saldo_var46',
       'delta_imp_amort_var18_1y3', 'delta_imp_amort_var34_1y3',
       'delta_imp_reemb_var17_1y3', 'delta_imp_ree

In [ ]:
# alternatively, we can remove the duplicated rows,
# transpose the dataframe back to the variables as columns
# keep first indicates that we keep the first of a set of
# duplicated variables

data_unique = data_t.drop_duplicates(keep='first').T
data_unique.shape

(10500, 265)

We can see immediately how removing duplicated features helps reduce the feature space. We passed from 370 to 265 non-duplicated features.

In [ ]:
# to find those columns in the original dataframe that were removed:

duplicated_features = [col for col in data.columns if col not in data_unique.columns]
duplicated_features 

['ind_var2',
 'ind_var13_medio_0',
 'ind_var13_medio',
 'ind_var18_0',
 'ind_var18',
 'ind_var26',
 'ind_var25',
 'ind_var27_0',
 'ind_var28_0',
 'ind_var28',
 'ind_var27',
 'ind_var29_0',
 'ind_var29',
 'ind_var32',
 'ind_var34_0',
 'ind_var34',
 'ind_var37',
 'ind_var40_0',
 'ind_var40',
 'ind_var41',
 'ind_var39',
 'ind_var44',
 'ind_var46_0',
 'ind_var46',
 'num_var13_medio_0',
 'num_var13_medio',
 'num_var18_0',
 'num_var18',
 'num_var26',
 'num_var25',
 'num_var27_0',
 'num_var28_0',
 'num_var28',
 'num_var27',
 'num_var29_0',
 'num_var29',
 'num_var32',
 'num_var34_0',
 'num_var34',
 'num_var37',
 'num_var40_0',
 'num_var40',
 'num_var41',
 'num_var39',
 'num_var46_0',
 'num_var46',
 'saldo_var13_medio',
 'saldo_var18',
 'saldo_var28',
 'saldo_var27',
 'saldo_var29',
 'saldo_var34',
 'saldo_var40',
 'saldo_var41',
 'saldo_var46',
 'delta_imp_amort_var18_1y3',
 'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var17_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_imp_trasp_var17_out_1y3

### Big datasets

Transposing a dataframe is memory costly if the dataframe is big. Therefore, we can use the alternative loop to find duplicated columns in bigger datasets.

In this case, I will use the same dataset, Santander from Kaggle, but I will load more rows. I expect to see less duplicated features, because by increasing the number of customers in the dataset, the probability of 2 customers having the same value across 2 or more features decreases. But this might as well not be the case. Let's have a look.

In [ ]:
# load the dataset
data = pd.read_csv('santander.csv', nrows=50000)

# separate into train and test sets
X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels=['TARGET'], axis=1),
    data['TARGET'],
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((35000, 370), (15000, 370))

In [ ]:
# check for duplicated features in the training set
duplicated_feat = []
for i in range(0, len(X_train.columns)):
    if i % 10 == 0:  # this helps me understand how the loop is going
        print(i)

    col_1 = X_train.columns[i]

    for col_2 in X_train.columns[i + 1:]:
        if X_train[col_1].equals(X_train[col_2]):
            duplicated_feat.append(col_2)

0
10
20
30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190
200
210
220
230
240
250
260
270
280
290
300
310
320
330
340
350
360


In [ ]:
# check how many features are duplicated
print(len(set(duplicated_feat)))

80


There are less duplicated features than when I loaded a smaller sample of the dataset. This behaviour is expected. Ideally you should work over the entire dataset.

In [ ]:
# let's print the list of duplicated features
set(duplicated_feat)

{'delta_imp_amort_var34_1y3',
 'delta_imp_reemb_var33_1y3',
 'delta_num_reemb_var13_1y3',
 'delta_num_reemb_var17_1y3',
 'delta_num_reemb_var33_1y3',
 'delta_num_trasp_var17_in_1y3',
 'delta_num_trasp_var17_out_1y3',
 'delta_num_trasp_var33_in_1y3',
 'delta_num_trasp_var33_out_1y3',
 'delta_num_venta_var44_1y3',
 'imp_amort_var18_hace3',
 'imp_amort_var34_hace3',
 'imp_reemb_var13_hace3',
 'imp_reemb_var17_hace3',
 'imp_reemb_var33_hace3',
 'imp_reemb_var33_ult1',
 'imp_trasp_var17_out_hace3',
 'imp_trasp_var33_out_hace3',
 'imp_venta_var44_hace3',
 'ind_var13_medio',
 'ind_var13_medio_0',
 'ind_var18',
 'ind_var2',
 'ind_var25',
 'ind_var26',
 'ind_var27',
 'ind_var27_0',
 'ind_var28',
 'ind_var28_0',
 'ind_var29',
 'ind_var29_0',
 'ind_var32',
 'ind_var34',
 'ind_var34_0',
 'ind_var37',
 'ind_var39',
 'ind_var41',
 'ind_var46',
 'ind_var46_0',
 'num_meses_var13_medio_ult3',
 'num_reemb_var13_hace3',
 'num_reemb_var17_hace3',
 'num_reemb_var33_hace3',
 'num_reemb_var33_ult1',
 'num_tr

In [ ]:
# we can go ahead and try to identify which set of features
# are identical

duplicated_feat = []
for i in range(0, len(X_train.columns)):

    col_1 = X_train.columns[i]

    for col_2 in X_train.columns[i + 1:]:

        # if the features are duplicated
        if X_train[col_1].equals(X_train[col_2]):

            #print them
            print(col_1)
            print(col_2)
            print()

            # and then append the duplicated one to a
            # list
            duplicated_feat.append(col_2)

ind_var2_0
ind_var2

ind_var2_0
ind_var13_medio_0

ind_var2_0
ind_var13_medio

ind_var2_0
ind_var27_0

ind_var2_0
ind_var28_0

ind_var2_0
ind_var28

ind_var2_0
ind_var27

ind_var2_0
ind_var34_0

ind_var2_0
ind_var34

ind_var2_0
ind_var41

ind_var2_0
ind_var46_0

ind_var2_0
ind_var46

ind_var2_0
num_var13_medio_0

ind_var2_0
num_var13_medio

ind_var2_0
num_var27_0

ind_var2_0
num_var28_0

ind_var2_0
num_var28

ind_var2_0
num_var27

ind_var2_0
num_var34_0

ind_var2_0
num_var34

ind_var2_0
num_var41

ind_var2_0
num_var46_0

ind_var2_0
num_var46

ind_var2_0
saldo_var13_medio

ind_var2_0
saldo_var28

ind_var2_0
saldo_var27

ind_var2_0
saldo_var34

ind_var2_0
saldo_var41

ind_var2_0
saldo_var46

ind_var2_0
delta_imp_amort_var34_1y3

ind_var2_0
delta_imp_reemb_var33_1y3

ind_var2_0
delta_num_reemb_var33_1y3

ind_var2_0
imp_amort_var18_hace3

ind_var2_0
imp_amort_var34_hace3

ind_var2_0
imp_reemb_var13_hace3

ind_var2_0
imp_reemb_var17_hace3

ind_var2_0
imp_reemb_var33_hace3

ind_var2_0
imp_re

In [ ]:
# let's check that indeed those features are duplicated
# I select a random pair from above

X_train[['ind_var2_0', 'num_var34_0']].head(10)

ind_var2_0  num_var34_0
17967           0            0
32391           0            0
9341            0            0
7929            0            0
46544           0            0
4149            0            0
33426           0            0
3002            0            0
6974            0            0
16864           0            0

In [ ]:
# let's check that indeed those features are duplicated
# I select another random pair from above

X_train[['ind_var2_0', 'ind_var2']].head(10)

ind_var2_0  ind_var2
17967           0         0
32391           0         0
9341            0         0
7929            0         0
46544           0         0
4149            0         0
33426           0         0
3002            0         0
6974            0         0
16864           0         0

We can see, that the features are identical.
